app/src/main/java/com/android/sample/helper/NotificationHelper.kt

```diff
@@ -0,0 +1,88 @@
+package com.android.sample.helper
+
+import android.app.AlarmManager
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.app.PendingIntent
+import android.content.Context
+import android.content.Intent
+import android.os.Build
+import android.util.Log
+import com.android.sample.receivers.NotificationReceiver
+import java.util.concurrent.TimeUnit
+
+class NotificationHelper(private val context: Context) {
+  private val notificationManager =
+      context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+  private val channelId = "activity_reminders"
+
+  init {
+    createNotificationChannel()
+  }
+
+  private fun createNotificationChannel() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      val channel =
+          NotificationChannel(
+                  channelId, "Activity Reminders", NotificationManager.IMPORTANCE_DEFAULT)
+              .apply { description = "Reminders for upcoming activities" }
+      notificationManager.createNotificationChannel(channel)
+    }
+  }
+
+  fun scheduleNotification(
+      activityId: Int,
+      activityName: String,
+      activityDate: Long,
+      notificationTitle: String
+  ) {
+    // cancel notification in case it already exists and just got updated
+    cancelNotification(activityId)
+    val intent =
+        Intent(context, NotificationReceiver::class.java).apply {
+          putExtra("activityId", activityId)
+          putExtra("activityName", activityName)
+          putExtra("notificationTitle", notificationTitle)
+        }
+
+    val pendingIntent =
+        PendingIntent.getBroadcast(
+            context,
+            activityId,
+            intent,
+            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
+
+    val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
+
+    val notificationTime = activityDate - TimeUnit.HOURS.toMillis(24) // 24h before
```

Comment:
As I also said in another comment, this will be ok for us because we just want to implement 1 more type of notification: a notification in the case the creator deletes an activity. In that case, I will be using Firebase Cloud Messaging (FCM) instead of AlarmManager (since it's event-based rather than time-based) and thus I can just keep scheduleNotification() as is and just add a new function fro the new task.

Reply:
Thank you for providing these details. I have this bad habit of wanting to optimise things, sometimes getting it a little out of scope of the project and vision.
But again, my concern is with the e2e, but yeah let's do it as you suggested. We'll go from there. Thanks !!

Reply:
On top of e2e, just to know but, testing becomes way easier if we use the appropriate structure.